src/controllers/user/register.js

```diff
@@ -12,8 +13,10 @@ const register = async (username, password) => {
     const newUser = new User();
     newUser.username = username;
     newUser.password = await User.generatePasswordHash(password);
+    const newCart = new Cart({ username: newUser.username });
 
     await newUser.save();
+    await newCart.save();
```

Comment:
Yup, user can only have cart. I am not sure why would a user have multiple cart... I dont shop online that much but I usually see one big cart where users can dump bunch of products in there.

Sample cart object:

```
cart = {
  username: 'antonio@gmail.com`,
  products = {
    productId1: {
      'S': 3
      'M': 4
    },
    productId2: {
      'S': 3
      'M': 4
    },
    productId3: {
      'L': 1
      'M': 1
    },
  }
}
```

S, M, L are the sizes of the product. They are used as key and their values indicate how many quantities for that selected size. How is this clear!

Reply:
Ok cool. It makes sense.

I'm seeing that retrieving the cart contents could however be tedious. Let's look at the following case:
* I want to know how many items are in a user's cart. Here you have 6 items, but there's no easy way to simple do a count programmatically. Based on your current structure you would have to do a recursive `lodash` forEach. Instead, consider the following structure:
```
cart = {
  username: 'antonio@gmail.com',
  products: [
    { id: productId1, quantity: 3, size: 'small', dateAdded: '2019-09-01' },
    { id: productId1, quantity: 4, size: 'medium' },
    { id: productId2, quantity: 3, size: 'small' },
    { id: productId2, quantity: 4 size: 'medium' },
    { id: productId3, quantity: 1, size: 'large' },
    { id: productId3, quantity: 1, size: 'medium' }
  ]
}
```
This way, you can simply do a `count` of the product array to know the total in your cart. You can also do `lodash.groupBy` or `lodash.sortBy` to group or sort by id or date or whatever field you need to. Basically, the goal is to have a versatile data structure that allows your frontend to easily transform the data in the way it needs.

Reply:
1. Yeah, I originally wanted to choose arrays but decided to just use objects. I am not a huge fan of arrays approach. Mainly because arrays have a linear time complexity for adding/removing vs objects/map having a constant constant time complexity for adding/removing https://medium.com/@sherryhsu/js-objects-and-arrays-which-one-is-faster-cfcdb1281704.

```
  products: [
    { id: productId1, quantity: 3, size: 'small', dateAdded: '2019-09-01' },
    { id: productId1, quantity: 4, size: 'medium' },
    { id: productId2, quantity: 3, size: 'small' },
    { id: productId2, quantity: 4 size: 'medium' },
    { id: productId3, quantity: 1, size: 'large' },
    { id: productId3, quantity: 1, size: 'medium' }
  ]
```

Suppose a user wants to update this item on their cart from
```    { id: productId3, quantity: 1, size: 'large' },```.

to

```    { id: productId3, quantity: 2, size: 'large' },```.
We would need to loop through the entire array which would take O(n) to find that item and then update the quantity field. But with objects, we don't need to loop and updating a field should take O(1).

_______
 
2. "Basically, the goal is to have a versatile data structure that allows your frontend to easily transform the data in the way it needs."

While this is true, we also need to consider the time complexity that some algorithms take in order to achieve what we want. If the cart arrays grow big, iterating through each items to remove/update an item will slow down our application.

Reply:
Having said that.... calculating the total number of items in the cart with my approach would take a long time due to having recursive calls so thats a big down for my approach. I'm unsure now about which approach is better. I will keep thinking about it :/

Reply:
1. Accessing properties in objects is in fact faster than in an array. However, that doesn't fit any of your use cases. t's good to understand theoretically what can happen but it only applies if in practice there is a real difference. There is only a margin difference in speed with the two approaches, which you will not see even at a large scale. Separately, I can't think of a case where you would need to select a single item (by id) from the cart.

2. The benefit here that I'm suggesting isn't to optimize for speed, it's to optimize for readability. Developers reading a recursive loop will find that it is more complex than a flat loop. So, code maintainability improves with the array approach as it is easier to handle the data when the structure is simpler.